Mathlib/Data/PEquiv.lean

```diff
@@ -0,0 +1,503 @@
+
+@[simp]
+theorem single_apply (a : α) (b : β) : single a b a = some b :=
+  if_pos rfl
+#align pequiv.single_apply PEquiv.single_apply
+
+theorem single_apply_of_ne {a₁ a₂ : α} (h : a₁ ≠ a₂) (b : β) : single a₁ b a₂ = none :=
+  if_neg h.symm
+#align pequiv.single_apply_of_ne PEquiv.single_apply_of_ne
+
+theorem single_trans_of_mem (a : α) {b : β} {c : γ} {f : β ≃. γ} (h : c ∈ f b) :
+    (single a b).trans f = single a c := by
+  ext
+  dsimp [single, PEquiv.trans]
+  split_ifs <;> simp_all
+#align pequiv.single_trans_of_mem PEquiv.single_trans_of_mem
+
+theorem trans_single_of_mem {a : α} {b : β} (c : γ) {f : α ≃. β} (h : b ∈ f a) :
+    f.trans (single b c) = single a c :=
+  symm_injective <| single_trans_of_mem _ ((mem_iff_mem f).2 h)
+#align pequiv.trans_single_of_mem PEquiv.trans_single_of_mem
+
+@[simp]
+theorem single_trans_single (a : α) (b : β) (c : γ) :
+    (single a b).trans (single b c) = single a c :=
+  single_trans_of_mem _ (mem_single _ _)
+#align pequiv.single_trans_single PEquiv.single_trans_single
+
+@[simp]
+theorem single_subsingleton_eq_refl [Subsingleton α] (a b : α) : single a b = PEquiv.refl α := by
+  ext (i j)
+  dsimp [single]
+  rw [if_pos (Subsingleton.elim i a), Subsingleton.elim i j, Subsingleton.elim b j]
+#align pequiv.single_subsingleton_eq_refl PEquiv.single_subsingleton_eq_refl
+
+theorem trans_single_of_eq_none {b : β} (c : γ) {f : δ ≃. β} (h : f.symm b = none) :
+    f.trans (single b c) = ⊥ := by
+  ext
+  simp only [eq_none_iff_forall_not_mem, Option.mem_def, f.eq_some_iff] at h
+  dsimp [PEquiv.trans, single]
+  simp
+  intros
+  split_ifs <;> simp_all
+#align pequiv.trans_single_of_eq_none PEquiv.trans_single_of_eq_none
+
+theorem single_trans_of_eq_none (a : α) {b : β} {f : β ≃. δ} (h : f b = none) :
+    (single a b).trans f = ⊥ :=
+  symm_injective <| trans_single_of_eq_none _ h
+#align pequiv.single_trans_of_eq_none PEquiv.single_trans_of_eq_none
+
+theorem single_trans_single_of_ne {b₁ b₂ : β} (h : b₁ ≠ b₂) (a : α) (c : γ) :
+    (single a b₁).trans (single b₂ c) = ⊥ :=
+  single_trans_of_eq_none _ (single_apply_of_ne h.symm _)
+#align pequiv.single_trans_single_of_ne PEquiv.single_trans_single_of_ne
+
+end Single
+
+section Order
+
+instance : PartialOrder
+      (α ≃. β) where
+  le f g := ∀ (a : α) (b : β), b ∈ f a → b ∈ g a
+  le_refl _ _ _ := id
+  le_trans f g h fg gh a b := gh a b ∘ fg a b
+  le_antisymm f g fg gf :=
+    ext
+      (by
+        intro a
+        cases' h : g a with b
+        · exact eq_none_iff_forall_not_mem.2 fun b hb => Option.not_mem_none b <| h ▸ fg a b hb
+        · exact gf _ _ h)
+
+theorem le_def {f g : α ≃. β} : f ≤ g ↔ ∀ (a : α) (b : β), b ∈ f a → b ∈ g a :=
+  Iff.rfl
+#align pequiv.le_def PEquiv.le_def
+
+instance : OrderBot (α ≃. β) :=
+  { instBotPEquiv with bot_le := fun _ _ _ h => (not_mem_none _ h).elim }
+
+instance [DecidableEq α] [DecidableEq β] : SemilatticeInf (α ≃. β) :=
+  { -- `split_ifs; finish` closes this goal from here
```

Comment:
This comment has mysteriously migrated: it's meant to be on the `split_ifs` below.